### ISSUE TITLE
Updating lead field ends with Mautic crash

### ISSUE THREAD

**Issue by pavel-janicek** [mautic/mautic#5821]
**Please DO NOT report security vulnerabilities here. Send them to security@mautic.com instead.**

What type of report is this:

| Q  | A
| ---| ---
| Bug report? | Y
| Feature request? | 
| Enhancement? | 

## Description:
GDPR hell
Create a campaign which checks on given segment if user visited certain page. If they do so, the system should update contact fields, namely if it is confirmed contact and if they want to be contacted. Later on, the campaign asks if such contact has value set and sends him campaign email. 

Campaign overview:
<img width="1101" alt="homeopatie_camaign_overall" src="https://user-images.githubusercontent.com/13867730/37558666-8e5d278a-2a17-11e8-9af6-40cf6fdf3bd4.png">

Visits a page decission:
<img width="593" alt="kampan_homeopatie_1" src="https://user-images.githubusercontent.com/13867730/37558671-a94b8c6c-2a17-11e8-9a58-c5c76f22d424.png">

Checking contact field:
<img width="602" alt="edit_campaign___mautic_homeoporadna" src="https://user-images.githubusercontent.com/13867730/37558676-bb1abc4c-2a17-11e8-85b9-8c32aaf520ae.png">

The page link is sent to the user via form which subscribes given user to an segment which is at top of this campaign and sends him email with content "click here if its you".

If user clicks that link - [link from mautic example](http://homeoporadna.eu/mautic/r/a33b17e31a7c83343f29f2ab8?ct=YTo1OntzOjY6InNvdXJjZSI7YToyOntpOjA7czo0OiJmb3JtIjtpOjE7aToxNTt9czo1OiJlbWFpbCI7aTo5MTtzOjQ6InN0YXQiO3M6MTM6IjVhYWQ1YmZhZDBjOGYiO3M6NDoibGVhZCI7aToxMjY5NztzOjc6ImNoYW5uZWwiO2E6MTp7czo1OiJlbWFpbCI7aTo5MTt9fQ%3D%3D&) they see Mautic Error 500 page instead of the landing page.

The system log is filled with this error:

```
Mar 17 10:23:30.305087 2018] [php7:error] [pid 20226] [client 2a01:7e01::f03c:91ff:fe9c:600e:49210] PHP Fatal error:  require(): Failed opening required '/var/www/html/homeoporadna.eu/public_html/mautic/app/cache/prod/doctrine/orm/Proxies/__CG__MauticUserBundleEntityUser.php' (include_path='.:/usr/share/php') in /var/www/html/homeoporadna.eu/public_html/mautic/vendor/doctrine/common/lib/Doctrine/Common/Proxy/AbstractProxyFactory.php on line 209
[Sat Mar 17 14:10:28.501645 2018] [php7:notice] [pid 23806] [client 157.55.39.69:5888] Doctrine\\ORM\\ORMInvalidArgumentException: A new entity was found through the relationship 'Mautic\\AssetBundle\\Entity\\Download#lead' that was not configured to cascade persist operations for entity: Mautic\\LeadBundle\\Entity\\Lead with ID #. To solve this issue: Either explicitly call EntityManager#persist() on this unknown entity or configure cascade persist  this association in the mapping for example @ManyToOne(..,cascade={"persist"}). in /var/www/html/homeoporadna.eu/public_html/mautic/vendor/doctrine/orm/lib/Doctrine/ORM/ORMInvalidArgumentException.php:92\nStack trace:\n#0 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php(852): Doctrine\\ORM\\ORMInvalidArgumentException::newEntityFoundThroughRelationship(Array, Object(Mautic\\LeadBundle\\Entity\\Lead))\n#1 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php(740): Doctrine\\ORM\\UnitOfWork->computeAssociationChanges(Array, Object(Mautic\\LeadBundle\\Entity\\Lead))\n#2 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php(452): Doctrine\\ORM\\UnitOfWork->computeChangeSet(Object(Doctrine\\ORM\\Mapping\\ClassMetadata), Object(Mautic\\AssetBundle\\Entity\\Download))\n#3 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php(485): Doctrine\\ORM\\UnitOfWork->computeScheduleInsertsChangeSets()\n#4 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php(342): Doctrine\\ORM\\UnitOfWork->computeSingleEntityChangeSet(Object(Mautic\\AssetBundle\\Entity\\Download))\n#5 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php(356): Doctrine\\ORM\\UnitOfWork->commit(Object(Mautic\\AssetBundle\\Entity\\Download))\n#6 /var/www/html/homeoporadna.eu/public_html/mautic/app/bundles/AssetBundle/Model/AssetModel.php(263): Doctrine\\ORM\\EntityManager->flush(Object(Mautic\\AssetBundle\\Entity\\Download))\n#7 /var/www/html/homeoporadna.eu/public_html/mautic/app/bundles/AssetBundle/Controller/PublicController.php(86): Mautic\\AssetBundle\\Model\\AssetModel->trackDownload(Object(Mautic\\AssetBundle\\Entity\\Asset), Object(Symfony\\Component\\HttpFoundation\\Request), 200)\n#8 [internal function]: Mautic\\AssetBundle\\Controller\\PublicController->downloadAction('3:pdf-o-solich')\n#9 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/symfony/http-kernel/HttpKernel.php(142): call_user_func_array(Array, Array)\n#10 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/symfony/http-kernel/HttpKernel.php(62): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)\n#11 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/symfony/http-kernel/DependencyInjection/ContainerAwareHttpKernel.php(67): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#12 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/symfony/http-kernel/Kernel.php(183): Symfony\\Component\\HttpKernel\\DependencyInjection\\ContainerAwareHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#13 /var/www/html/homeoporadna.eu/public_html/mautic/app/AppKernel.php(133): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#14 /var/www/html/homeoporadna.eu/public_html/mautic/app/middlewares/CORSMiddleware.php(93): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#15 /var/www/html/homeoporadna.eu/public_html/mautic/app/middlewares/CatchExceptionMiddleware.php(45): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#16 /var/www/html/homeoporadna.eu/public_html/mautic/app/middlewares/VersionCheckMiddleware.php(57): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#17 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#18 /var/www/html/homeoporadna.eu/public_html/mautic/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))\n#19 /var/www/html/homeoporadna.eu/public_html/mautic/index.php(34): Stack\\run(Object(Stack\\StackedHttpKernel))\n#20 {main}
[Sat Mar 17 14:41:21.251972 2018] [access_compat:error] [pid 24305] [client 163.172.255.139:57248] AH01797: client denied by server configuration: /var/www/html/homeoporadna.eu/public_html/wp-includes/css/modules.php, referer: http://homeoporadna.eu/wp-includes/css/modules.php)
```
What can I do to avoid this bug?

Very strange thing is that if an user clicks the same link for second time, they are correctly redirected to confirmation page

## If a bug:

| Q   | A
| --- | ---
| Mautic version |  2.12.2
| PHP version | PHP 7.1.11-0ubuntu0.17.10.1 (cli) (built: Nov  1 2017 16:30:52) ( NTS )

### Steps to reproduce:
1. Set up a new email using mailinator.com
2. Go to http://homeoporadna.eu/homeopatie-zdarma/ in incognito window
3. Fill in email obtined on mailinator.com to form present on a page
4. Go to mailinator inbox
5. click the first green link
 
### Log errors: 
see above
_Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable._

**Comment by pavel-janicek** [comment]
Also, what is funny, all is updated on the background. So when user tries to visit the page, mautic recognizes it as a page hit, updates the contact and sends the e-book. It also triggers campaign correctly. Only thing which seems to be wrong is, that first page click ends with an error page instead of correct landing page on the web

**Comment by pavel-janicek** [comment]
Possibly related to #5391

**Comment by pavel-janicek** [comment]
Can be partially solved by #5678 but the contact is then doubled in contact list

**Comment by pavel-janicek** [comment]
**SOLVED!**
The problem was in campaign itself. Basically I was trying to read from field which was just being set by campaign itself. Big thanks to my non-programmer wife who was able to spot my error in campaign logic.

This is how the campaign looks now
<img width="1231" alt="homeoporadna_correct_campaign" src="https://user-images.githubusercontent.com/13867730/37560106-9de9a230-2a32-11e8-9734-6a6b264385ad.png">
